### PR TITLE
Enable alternative types for certain fields when decoding a `ProductVariation`

### LINKED ISSUE #2401: [Product Details] [Variations] Can't see the variations list when selecting "Variations"
On the product details screen, when I select the variation row, I get an error message and can't see the variation list.

https://cloudup.com/co4_r6oy1DS

To reproduce:
1. Open Product details screen for a products with variations. 
2. Select "Variations"

### PR DESCRIPTION
**Pull request by jaclync**
Fixes #2401 

Note: ~304 diffs of this PR are from JSON files used to test decoding a product or product variation in different edge cases!

In addition to fixing the issue when a site's `ProductVariation` might have different types from expected, this PR also includes some refactoring on how alternative types are decoded.

## Changes

All the changes are in the Networking layer:

- Created `AlternativeDecodingType` enum with 3 types that are currently used:
  - `decimal`: used to try decoding a value of `Decimal` type
  - `string`: used to try decoding a value of `String` type
  - `bool`: used to try decoding a value of `Bool` type
- In `KeyedDecodingContainer`'s extension, added another helper `failsafeDecodeIfPresent` that takes a target type for the final type with a list of alternative types with how each alternative type is transformed to the target type
- In `Product`, updated the alternative types conditional logic with the new `failsafeDecodeIfPresent` helper with alternative types
  - Right now, we support alternative types for `price`, `salePrice`, and `manageStock`
  - Added a test case with a JSON that contains the fields with alternative types
- In `ProductVariation`, added/updated the alternative types conditional logic with the new `failsafeDecodeIfPresent` helper with alternative types
  - Similar to `Product`, we support alternative types for `price`, `salePrice`, and `manageStock`
  - Added test cases with JSON files for decoding a list of `ProductVariation`s with `ProductVariationListMapper`

## Testing

Fortunately, I was able to repro this on @Garance91540's site (thanks to Rachel's investigation in p5T066-1o1#comment-5400-p2)! There is probably a third-party plugin that alters the price type from string to decimal. However, feel free to test this on any site.

- Launch the app
- Go to the products tab --> the product list should load as before
- Tap on a variable product
- Tap on the variations row --> the product variation list should load correctly

## Example screenshots

`develop` | PR branch
-- | --
Simulator Screen Shot - iPhone 11 - 2020-07-17 at 18 26 10 | Simulator Screen Shot - iPhone 11 - 2020-07-17 at 18 28 05

Update release notes:
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: Networking/Networking/Extensions/KeyedDecodingContainer+Woo.swift [hunk 1]
```diff
@@ -18,6 +24,42 @@ extension KeyedDecodingContainer {
...
+    /// - Parameters:
+    ///   - targetType: target type for a given key.
+    ///   - key: the key that maps to the value to be decoded.
+    ///   - shouldDecodeTargetTypeFirst: whether it should try decoding to the target type without going through the alternative types first.
+    ///   - alternativeTypes: a list of alternative types that can be mapped to the target type.
+    /// - Returns: optional value of the target type.
+    func failsafeDecodeIfPresent<T>(targetType: T.Type,
+                                    forKey key: KeyedDecodingContainer<K>.Key,
+                                    shouldDecodeTargetTypeFirst: Bool = true,
+                                    alternativeTypes: [AlternativeDecodingType<T>]) -> T? where T: Decodable {
```
Comment (pmusolino) [inline Networking/Networking/Extensions/KeyedDecodingContainer+Woo.swift:37]: This is Amazing! 💯

File: Networking/NetworkingTests/Mapper/ProductMapperTests.swift [hunk 2]
```diff
@@ -105,6 +101,17 @@ class ProductMapperTests: XCTestCase {
...
+    /// Verifies that the fields of the Product with alternative types are parsed correctly when they have different types than in the struct.
+    /// Currently, `price`, `salePrice` and `manageStock` allow alternative types.
+    ///
+    func testThatProductAlternativeTypesAreProperlyParsed() throws {
+        let product = try XCTUnwrap(mapLoadProductResponseWithAlternativeTypes())
+
+        XCTAssertEqual(product.price, "17")
+        XCTAssertEqual(product.salePrice, "26.73")
+        XCTAssertTrue(product.manageStock)
+    }
```
Comment (pmusolino) [inline Networking/NetworkingTests/Mapper/ProductMapperTests.swift:113]: 🥇
